File: flutter_readium/android/src/main/kotlin/dk/nota/flutter_readium/ReadiumReaderChannel.kt
package dk.nota.flutter_readium

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import org.json.JSONObject
import org.readium.r2.shared.publication.Locator
import org.readium.r2.shared.util.AbsoluteUrl

internal class ReadiumReaderChannel(
    messenger: BinaryMessenger,
    name: String,
) : MethodChannel(messenger, name),
    CoroutineScope by CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate) {
    fun onPageChanged(locator: Locator?) =
        launch {
            invokeMethod(
                "onPageChanged",
                locator?.toJSON().toString(),
            )
        }

    fun onExternalLinkActivated(url: AbsoluteUrl) = launch { invokeMethod("onExternalLinkActivated", url.toString()) }

    fun onImageTapped(href: String) = launch { invokeMethod("onImageTapped", href) }

    fun onTextSelected(locator: Locator, selectedText: String?) =
        launch {
            val json = JSONObject().apply {
                put("locator", locator.toJSON())
                put("selectedText", selectedText ?: JSONObject.NULL)
            }
            invokeMethod("onTextSelected", json.toString())
        }

    fun onSelectionAction(actionId: String, locator: Locator, selectedText: String?) =
        launch {
            val json = JSONObject().apply {
                put("actionId", actionId)
                put("locator", locator.toJSON())
                put("selectedText", selectedText ?: JSONObject.NULL)
            }
            invokeMethod("onSelectionAction", json.toString())
        }

    fun onDecorationInteraction(decorationId: String, group: String, type: String, locator: Locator?) =
        launch {
            val json = JSONObject().apply {
                put("decorationId", decorationId)
                put("group", group)
                put("type", type)
                if (locator != null) put("locator", locator.toJSON()) else put("locator", JSONObject.NULL)
            }
            invokeMethod("onDecorationInteraction", json.toString())
        }
}

File: flutter_readium/ios/flutter_readium/Sources/flutter_readium/ReadiumReaderChannel.swift
import Flutter
import ReadiumShared

class ReadiumReaderChannel: FlutterMethodChannel {
  // Compiles fine without this init, but then mysteriously crashes later with EXC_BAD_ACCESS when calling any member function later.
  init(name: String, binaryMessenger messenger: FlutterBinaryMessenger) {
    super.init(
      name: name, binaryMessenger: messenger, codec: FlutterStandardMethodCodec.sharedInstance(),
      taskQueue: nil)
  }

  func onPageChanged(locator: Locator) {
    invokeMethod("onPageChanged", arguments: try? locator.jsonString())
  }

  func onExternalLinkActivated(url: URL) {
    invokeMethod("onExternalLinkActivated", arguments: url.absoluteString as String?)
  }

  func onImageTapped(href: String) {
    invokeMethod("onImageTapped", arguments: href)
  }

  func onTextSelected(locator: Locator, selectedText: String?) {
    var json: [String: Any] = [:]
    if let locatorString = try? locator.jsonString(),
       let locatorData = locatorString.data(using: .utf8),
       let locatorJson = try? JSONSerialization.jsonObject(with: locatorData) {
      json["locator"] = locatorJson
    }
    if let selectedText = selectedText {
      json["selectedText"] = selectedText
    }
    if let data = try? JSONSerialization.data(withJSONObject: json),
       let jsonString = String(data: data, encoding: .utf8) {
      invokeMethod("onTextSelected", arguments: jsonString)
    }
  }

  func onSelectionAction(actionId: String, locator: Locator, selectedText: String?) {
    var json: [String: Any] = ["actionId": actionId]
    if let locatorString = try? locator.jsonString(),
       let locatorData = locatorString.data(using: .utf8),
       let locatorJson = try? JSONSerialization.jsonObject(with: locatorData) {
      json["locator"] = locatorJson
    }
    if let selectedText = selectedText {
      json["selectedText"] = selectedText
    }
    if let data = try? JSONSerialization.data(withJSONObject: json),
       let jsonString = String(data: data, encoding: .utf8) {
      invokeMethod("onSelectionAction", arguments: jsonString)
    }
  }

  func onDecorationInteraction(decorationId: String, group: String, type: String, locator: Locator?) {
    var json: [String: Any] = [
      "decorationId": decorationId,
      "group": group,
      "type": type,
    ]
    if let locator = locator,
       let locatorString = try? locator.jsonString(),
       let locatorData = locatorString.data(using: .utf8),
       let locatorJson = try? JSONSerialization.jsonObject(with: locatorData) {
      json["locator"] = locatorJson
    }
    if let data = try? JSONSerialization.data(withJSONObject: json),
       let jsonString = String(data: data, encoding: .utf8) {
      invokeMethod("onDecorationInteraction", arguments: jsonString)
    }
  }
}
